### PR TITLE
P4-2170 changes to tighten and improve the validation around the locations import process.

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
These changes tighten up the location validation steps when importing the locations spreadsheet.

Persistence level exceptions have been reduced as we are now checking prior to save the locations entity in the DB.

The locations import will also fail (fast) if any of the expected location spreadsheet tabs are missing.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/location/importer/LocationsSpreadsheet.kt [hunk 1]
```diff
@@ -36,11 +43,21 @@ class LocationsSpreadsheet(private val spreadsheet: Workbook) : Closeable {
     fun getRowsFrom(tab: Tab): List<Row> = spreadsheet.getSheet(tab.label).drop(COLUMN_HEADINGS).filterNot { it.getCell(1)?.stringCellValue.isNullOrBlank() }
 
     fun mapToLocation(cells: List<Cell>): Location {
+        val locationType = LocationType.map(cells[TYPE].stringCellValue.toUpperCase().trim())
+                ?: throw IllegalArgumentException("Unsupported location type: " + cells[TYPE].stringCellValue)
+
+        val agency = cells[AGENCY].stringCellValue.toUpperCase().trim().takeUnless { it == "" }
```
Comment (jcavell) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/location/importer/LocationsSpreadsheet.kt:49]: isBlank() might be better

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/location/importer/LocationsSpreadsheet.kt [hunk 2]
```diff
@@ -4,19 +4,26 @@ import org.apache.poi.ss.usermodel.Cell
...
 
 /**
  * Simple wrapper class to encapsulate the logic around access to data in the locations spreadsheet. When finished with the spreadsheet should be closed.
  */
-class LocationsSpreadsheet(private val spreadsheet: Workbook) : Closeable {
+class LocationsSpreadsheet(private val spreadsheet: Workbook, private val locationRepository: LocationRepository) : Closeable {
+
+    init {
+        Tab.values().toList().filter { spreadsheet.getSheet(it.label) == null }.joinToString { it.label }.let {
+            if (it.isNotBlank()) throw IllegalStateException("The following tabs are missing from the locations spreadsheet: $it")
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/location/importer/LocationsSpreadsheet.kt:24]: I think you can do `.takeIf{it.IsNotBlank()}.let{throw Illegal ....}`

or some such syntax

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/location/importer/LocationsSpreadsheet.kt (hunk 1 above)
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/location/importer/LocationsSpreadsheet.kt:49]:
```suggestion
        val agency = cells[AGENCY].stringCellValue.toUpperCase().trim().takeUnless { it == "" }!!
```

if you are finding the code a bit noisy you could force a NullPointerException

**Review by jcavell — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/location/importer/LocationsSpreadsheet.kt (hunk 1 above)
Comment (andymarke) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/location/importer/LocationsSpreadsheet.kt:49]: Only makes sense if you don't care what exception you get

**Review by andymarke — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/location/importer/LocationsSpreadsheet.kt (hunk 1 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/location/importer/LocationsSpreadsheet.kt:49]: Indeed it would.  Will address in future Pr.

Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/location/importer/LocationsSpreadsheet.kt:49]: Hmm don't think appropriate but does raise the question of the exception as could be an NPE instead.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/location/importer/LocationsSpreadsheet.kt (hunk 2 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/location/importer/LocationsSpreadsheet.kt:24]: Will `takeIf { it.works } `....I'll get my coat.

**Pull request merged**